pkg/controllers/nodeipam/ipam/cidr_allocator.go

```diff
@@ -0,0 +1,480 @@
+	// We must poll because apiserver might not be up. This error causes
+	// controller manager to restart.
+	if pollErr := wait.Poll(nodePollInterval, apiserverStartupGracePeriod, func() (bool, error) {
+		var err error
+		nodeList, err = kubeClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
+			FieldSelector: fields.Everything().String(),
+			LabelSelector: labels.Everything().String(),
+		})
+		if err != nil {
+			klog.Errorf("Failed to list all nodes: %v", err)
+			return false, nil
+		}
+		return true, nil
+	}); pollErr != nil {
+		return nil, fmt.Errorf("failed to list all nodes in %v, cannot proceed without updating CIDR map",
+			apiserverStartupGracePeriod)
+	}
+	return nodeList, nil
+}
+
+type rangeAllocator struct {
+	client clientset.Interface
+	// cluster cidrs as passed in during controller creation
+	clusterCIDRs []*net.IPNet
+	// for each entry in clusterCIDRs we maintain a list of what is used and what is not
+	cidrSets []*cidrset.CidrSet
+	// nodeLister is able to list/get nodes and is populated by the shared informer passed to controller
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	nodesSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan NodeReservedCIDRs
+	recorder              record.EventRecorder
+	// Keep a set of nodes that are currently being processed to avoid races in CIDR allocation
+	lock              sync.Mutex
+	nodesInProcessing sets.String
+	cloud             *awsv1.Cloud
+}
+
+// NewCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one from each of clusterCIDRs)
+// Caller must ensure subNetMaskSize is not less than cluster CIDR mask size.
+// Caller must always pass in a list of existing nodes so the new allocator.
+// can initialize its CIDR map. NodeList is only nil in testing.
+func NewCIDRRangeAllocator(client clientset.Interface, nodeInformer informers.NodeInformer, awsCloud *awsv1.Cloud, allocatorParams CIDRAllocatorParams, nodeList *v1.NodeList) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("kubeClient is nil when starting NodeController")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cidrAllocator"})
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
+
+	// create a cidrSet for each CIDR we operate on.
+	// cidrSet are mapped to clusterCIDR by index
+	cidrSets := make([]*cidrset.CidrSet, len(allocatorParams.ClusterCIDRs))
+	for idx, cidr := range allocatorParams.ClusterCIDRs {
+		cidrSet, err := cidrset.NewCIDRSet(cidr, allocatorParams.NodeCIDRMaskSizes[idx])
```

Comment:
nit, this can cause crash if the len(allocatorParams.NodeCIDRMaskSizes) <  len(allocatorParams.ClusterCIDRs).
It's a nit comment given this shall be supplied by cluster admins so lt's low risk, but good to have validations with clear error message when len(allocatorParams.NodeCIDRMaskSizes) != len(allocatorParams.ClusterCIDRs)